bug59153_ExceptionInCSVDataSetWithEmptyFilename

ra0077 opened the pull request:
Hi,

When you add a CSV Data Set Config in your test plan and you don't fill
the filename field we have an exception

ERROR - jmeter.threads.JMeterThread: Test failed!
java.lang.IllegalArgumentException: Could not read file header line
    at
org.apache.jmeter.services.FileServer.reserveFile(FileServer.java:280)
    at
org.apache.jmeter.config.CSVDataSet.iterationStart(CSVDataSet.java:178)
    at
org.apache.jmeter.control.GenericController.fireIterationStart(GenericController.java:405)
    at
org.apache.jmeter.control.GenericController.fireIterEvents(GenericController.java:397)
    at
org.apache.jmeter.control.GenericController.next(GenericController.java:158)
    at
org.apache.jmeter.control.LoopController.next(LoopController.java:123)
    at
org.apache.jmeter.threads.AbstractThreadGroup.next(AbstractThreadGroup.java:87)
    at org.apache.jmeter.threads.JMeterThread.run(JMeterThread.java:238)
    at java.lang.Thread.run(Unknown Source)
Caused by: java.io.FileNotFoundException: C:\XXXXXXXXX (Accès refusé)
    at java.io.FileInputStream.open0(Native Method)
    at java.io.FileInputStream.open(Unknown Source)
    at java.io.FileInputStream.<init>(Unknown Source)
    at
org.apache.jmeter.services.FileServer.createBufferedReader(FileServer.java:419)
    at org.apache.jmeter.services.FileServer.readLine(FileServer.java:345)
    at org.apache.jmeter.services.FileServer.readLine(FileServer.java:329)
    at
org.apache.jmeter.services.FileServer.reserveFile(FileServer.java:277)
    ... 8 more

Antonio

ra0077 committed:
Hi,

Whe you add a CSV Data Set Config in your test plan and you don't fill
the filename field we have an exception

ERROR - jmeter.threads.JMeterThread: Test failed!
java.lang.IllegalArgumentException: Could not read file header line
	at
org.apache.jmeter.services.FileServer.reserveFile(FileServer.java:280)
	at
org.apache.jmeter.config.CSVDataSet.iterationStart(CSVDataSet.java:178)
	at
org.apache.jmeter.control.GenericController.fireIterationStart(GenericController.java:405)
	at
org.apache.jmeter.control.GenericController.fireIterEvents(GenericController.java:397)
	at
org.apache.jmeter.control.GenericController.next(GenericController.java:158)
	at
org.apache.jmeter.control.LoopController.next(LoopController.java:123)
	at
org.apache.jmeter.threads.AbstractThreadGroup.next(AbstractThreadGroup.java:87)
	at org.apache.jmeter.threads.JMeterThread.run(JMeterThread.java:238)
	at java.lang.Thread.run(Unknown Source)
Caused by: java.io.FileNotFoundException: C:\XXXXXXXXX (Accès refusé)
	at java.io.FileInputStream.open0(Native Method)
	at java.io.FileInputStream.open(Unknown Source)
	at java.io.FileInputStream.<init>(Unknown Source)
	at
org.apache.jmeter.services.FileServer.createBufferedReader(FileServer.java:419)
	at org.apache.jmeter.services.FileServer.readLine(FileServer.java:345)
	at org.apache.jmeter.services.FileServer.readLine(FileServer.java:329)
	at
org.apache.jmeter.services.FileServer.reserveFile(FileServer.java:277)
	... 8 more

Antonio

FSchumacher commented:
Thanks for your contribution.

I have two comments.

The formatting seems to be broken by this change.
I think jmeter should not skip the missing filename silently.

ra0077 commented:
Hi,

I can fix the formatting.

About the second comments, what do you want I do? 
Do I need to log it in debug?

Antonio

ra0077 committed:
bug59153_ExceptionInCSVDataSetWithEmptyFilename/PR162
Fix the formatting
JMeter should not skip the missing filename silently

ra0077 commented:
New commit

Formatting fixed
I have add a warning message (log.warn) to not skip the missing filename silently

Antonio

pmouawad commented:
Hi Antonio,
If file name is empty what is the use of CSV DataSet ?

I think maybe we should make the test fail and stop , which does not appear to be the case currently.
Thanks
Regards

ra0077 commented:
Hi Philippe,

I was training a colleague to JMeter and I have add a CSV DataSet to use it later in the test plan.
I run the test to check it before fill the CSV DataSet and I have this exception
And it's not the first time I have it (when I work wit people who are not expert I like to add all the element I want he will use it without fill all of them)

To answer your question, I add a CSV DataSet with empty file name to:
  tell to someone to use it
  use it later when my test plan will be checked

With my use, the test will not failed and stop but continue (maybe with a warning like I have implemented)

In all cases the need to ckeck the log to see the exception to understand why the script don't start is not a good solution

Regards
Antonio

ra0077 committed:
bug59153_ExceptionInCSVDataSetWithEmptyFilename

Hi,

JMeter not skip the missing filename silently (like previously)

But now instead of having an FileNotFoundException exception +
stacktrace we have a clear error message + JMeterStopThreadException

Antonio

ra0077 commented:
Hi,

JMeter not skip the missing filename silently (like previously)

But now instead of having an FileNotFoundException exception + stacktrace we have a clear error message + JMeterStopThreadException

Antonio

vlsi commented on src/components/org/apache/jmeter/config/CSVDataSet.java:
```diff
@@ -146,71 +146,76 @@ public void setProperty(JMeterProperty property) {
+                        alias = _fileName;
+                        break;
+                    case CSVDataSetBeanInfo.SHARE_GROUP:
+                        alias = _fileName+"@"+System.identityHashCode(context.getThreadGroup());
```
Do you assume "all the thread groups would get unique hash codes". I'm afraid that would not work.

vlsi commented on src/components/org/apache/jmeter/config/CSVDataSet.java:
```diff
@@ -146,71 +146,76 @@ public void setProperty(JMeterProperty property) {
+                        alias = _fileName+"@"+System.identityHashCode(context.getThreadGroup());
+                        break;
+                    case CSVDataSetBeanInfo.SHARE_THREAD:
+                        alias = _fileName+"@"+System.identityHashCode(context.getThread());
```
Same for `..identityHashCode(context.getThread...`

ra0077 commented:
Hi,

I don't have modify these lines.

But I think that CSV Data Set need more work to be ok

Antonio

vlsi commented on src/components/org/apache/jmeter/config/CSVDataSet.java:
```diff
@@ -146,71 +146,76 @@ public void setProperty(JMeterProperty property) {
             }
+        } else {
+            log.error("No filename setup in CSV Data Set Config: "+this.getName());
+            throw new JMeterStopThreadException("No filename setup in CSV Data Set Config: "+this.getName());
```
Please swap order of those branches. It is hard to follow where is the `if`

vlsi commented:
>I don't have modify these lines.

@ra0077, can you please rework the PR without reformatting the whole method?

ra0077 commented:
I have swap it and remove formating to be easy to read it

What I do is:
check if the user have put a file name
if not
log it to allow easy debuging
stop the test
if yes
no modification


It allow to easy debug (a clear message in log) and avoid to have an
exception to user



The diff with the trunk:

--- a/src/components/org/apache/jmeter/config/CSVDataSet.java
+++ b/src/components/org/apache/jmeter/config/CSVDataSet.java
@@ -146,6 +146,11 @@ public class CSVDataSet extends ConfigTestElement

     @Override
     public void iterationStart(LoopIterationEvent iterEvent) {
+        String _fileName = getFilename();
+        if (_fileName.isEmpty()) {
+            log.error("No filename setup in CSV Data Set Config:
"+this.getName());
+            throw new JMeterStopThreadException("No filename setup in CSV
Data Set Config: "+this.getName());
+        } else {
         FileServer server = FileServer.getFileServer();
         final JMeterContext context = getThreadContext();
         String delim = getDelimiter();
@@ -156,7 +161,6 @@ public class CSVDataSet extends ConfigTestElement
             delim=",";
         }
         if (vars == null) {
-            String _fileName = getFilename();
             String mode = getShareMode();
             int modeInt = CSVDataSetBeanInfo.getShareModeAsInt(mode);
             switch(modeInt){
@@ -212,6 +216,7 @@ public class CSVDataSet extends ConfigTestElement
                 threadVars.put(var, EOFVALUE);
             }
         }
+        }
     }

Antonio

Cet e-mail a été envoyé depuis un ordinateur protégé par Avast.
www.avast.com
<https://www.avast.com/fr-fr/lp-safe-emailing?utm_medium=email&utm_source=link&utm_campaign=sig-email&utm_content=webmail&utm_term=OA-2109-A>
<#DDB4FAA8-2DD7-40BB-A1B8-4E2AA1F9FDF2>

2016-03-15 9:13 GMT+01:00 Vladimir Sitnikov <notifications@github.com>:

> I don't have modify these lines.
>
> @ra0077 <https://github.com/ra0077>, can you please rework the PR without
> reformatting the whole method?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub:
> https://github.com/apache/jmeter/pull/162#issuecomment-196710977
>

ra0077 commented:
I have swap it and remove formating to be easy to read it

What I do is:
check if the user have put a file name
if not
  log it to allow easy debuging
  stop the test
if yes
  no modification


It allow to easy debug (a clear message in log) and avoid to have an exception to user



The diff with the trunk:

--- a/src/components/org/apache/jmeter/config/CSVDataSet.java
+++ b/src/components/org/apache/jmeter/config/CSVDataSet.java
@@ -146,6 +146,11 @@ public class CSVDataSet extends ConfigTestElement
 
     @Override
     public void iterationStart(LoopIterationEvent iterEvent) {
+        String _fileName = getFilename();
+        if (_fileName.isEmpty()) {
+            log.error("No filename setup in CSV Data Set Config: "+this.getName());
+            throw new JMeterStopThreadException("No filename setup in CSV Data Set Config: "+this.getName());
+        } else {
         FileServer server = FileServer.getFileServer();
         final JMeterContext context = getThreadContext();
         String delim = getDelimiter();
@@ -156,7 +161,6 @@ public class CSVDataSet extends ConfigTestElement
             delim=",";
         }
         if (vars == null) {
-            String _fileName = getFilename();
             String mode = getShareMode();
             int modeInt = CSVDataSetBeanInfo.getShareModeAsInt(mode);
             switch(modeInt){
@@ -212,6 +216,7 @@ public class CSVDataSet extends ConfigTestElement
                 threadVars.put(var, EOFVALUE);
             }
         }
+        }
     }

vlsi commented:
>I have swap it and remove formating to be easy to read it

@ra0077, would you please update the PR then?

ra0077 commented:
New PR 167 submittted
